custom_components/battery_sim/__init__.py

```diff
@@ -393,79 +483,122 @@ def getTariffReading(self, entity_id):
+        amount_to_charge: float = 0.0
+        amount_to_discharge: float = 0.0
+        net_export: float = 0.0
+        net_import: float = 0.0
+
         if self._charge_state == "unknown":
             self._charge_state = 0.0
 
-        """Calculate maximum possible charge and discharge based on battery specifications and time since last discharge"""
+        """
+            Calculate maximum possible charge and discharge based on battery
+            specifications and time since last discharge
+        """
         time_now = time.time()
-        time_since_last_battery_update = time_now - self._last_battery_update_time
-        max_discharge = time_since_last_battery_update * self._max_discharge_rate / 3600
-        max_charge = time_since_last_battery_update * self._max_charge_rate / 3600
-        available_capacity_to_charge = self._battery_size - float(self._charge_state)
+        time_last_update = self._last_battery_update_time
+        time_since_last_battery_update = time_now - time_last_update
+
+        _LOGGER.debug(
+            "(%s) Import: (%s) Export: (%s) .... Timing: %s = Now / %s = Last Update / %s Time (sec).",
+            self._name,
+            import_amount,
+            export_amount,
+            time_now,
+            time_last_update,
+            time_since_last_battery_update,
+        )
+
+        max_discharge = time_since_last_battery_update * (
+            self._max_discharge_rate / 3600
+        )
+        max_charge = time_since_last_battery_update * (
+            self._max_charge_rate / 3600
+        )
+
+        available_capacity_to_charge = (
+            self._battery_size - float(self._charge_state)
+        )
+
         available_capacity_to_discharge = float(self._charge_state) * float(
             self._battery_efficiency
         )
 
+        if not any(self._switches.values()):
+            _LOGGER.debug("(%s) Battery normal mode.", self._name)
+
+            amount_to_charge = min(
+                export_amount, max_charge, available_capacity_to_charge
+            )
+            amount_to_discharge = min(
+                import_amount, max_discharge, available_capacity_to_discharge
+            )
+            net_import = import_amount - amount_to_discharge
+            net_export = export_amount - amount_to_charge
+            if amount_to_charge > amount_to_discharge:
+                self._sensors[BATTERY_MODE] = MODE_CHARGING
+            else:
+                self._sensors[BATTERY_MODE] = MODE_DISCHARGING
+
         if self._switches[PAUSE_BATTERY]:
-            _LOGGER.debug("Battery (%s) paused.", self._name)
+            _LOGGER.debug("(%s) Battery paused.", self._name)
             amount_to_charge = 0.0
             amount_to_discharge = 0.0
             net_export = export_amount
             net_import = import_amount
             self._sensors[BATTERY_MODE] = MODE_IDLE
-        elif self._switches[OVERIDE_CHARGING]:
```

Comment:
By making these separate if statements not elif what will happen if multiple switches are on. e.g. if the pause battery and the force discharge are both on then the battery will do whatever is the last action in the chain I guess as long as all the variables are updated in each if block. I do like having the default action first because that is easier to read. I think pause should take precedence over the other options though so maybe move that to the end or reinstate the elif. At least the elif makes it clear that an exclusive choice is being made. The other option would be to somehow link the buttons so only one can be on, but then you couldn't pause and unpause without somehow remembering the previous state. What do you think?